Sources/ServiceDiscovery/ServiceDiscovery+AsyncAwait.swift

```diff
@@ -0,0 +1,91 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Dispatch
+
+#if compiler(>=5.5) && canImport(_Concurrency)
+
+public extension ServiceDiscovery {
+    @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+    func lookup(_ service: Service, deadline: DispatchTime? = nil) async throws -> [Instance] {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<[Instance], Error>) in
+            self.lookup(service, deadline: deadline) { result in
+                switch result {
+                case .success(let instances):
+                    continuation.resume(returning: instances)
+                case .failure(let error):
+                    continuation.resume(throwing: error)
+                }
+            }
+        }
+    }
+
+    @available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+    func subscribe(to service: Service) -> ServiceSnapshots<Instance> {
+        ServiceSnapshots(AsyncThrowingStream<[Instance], Error> { continuation in
+            Task {
+                let cancellationToken = self.subscribe(
+                    to: service,
+                    onNext: { result in
+                        switch result {
+                        case .success(let instances):
+                            continuation.yield(instances)
+                        case .failure(let error):
+                            // LookupError is recoverable (e.g., service is added *after* subscription begins), so don't give up yet
+                            guard error is LookupError else {
+                                return continuation.finish(throwing: error)
+                            }
+                        }
+                    },
+                    onComplete: { reason in
+                        switch reason {
+                        case .cancellationRequested:
+                            continuation.finish()
+                        case .serviceDiscoveryUnavailable:
+                            continuation.finish(throwing: ServiceDiscoveryError.unavailable)
+                        default:
+                            continuation.finish(throwing: ServiceDiscoveryError.other(reason.description))
+                        }
+                    }
+                )
+
+                continuation.onTermination = { @Sendable (_) -> Void in
+                    cancellationToken.cancel()
+                }
+            }
+        })
+    }
+}
+
+@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+public struct ServiceSnapshots<Instance>: AsyncSequence, AsyncIteratorProtocol {
+    public typealias Element = [Instance]
+
+    private let _next: () async throws -> [Instance]?
+
+    public init<SnapshotSequence: AsyncSequence>(_ snapshots: SnapshotSequence) where SnapshotSequence.Element == Element {
+        var iterator = snapshots.makeAsyncIterator()
+        self._next = { try await iterator.next() }
```

Comment:
Do we need the closure? If our iterator just stores the underlying iterator, I think we can just call `next` on our instance of it.